include/AdePT/core/AdePTTransport.cuh

```diff
@@ -1,100 +1,458 @@
-  MParrayTracks *leakedPositrons;
-  MParrayTracks *leakedGammas;
-};
-
-struct ParticleType {
-  adept::TrackManager<Track> *trackmgr;
-  MParrayTracks *leakedTracks;
-  cudaStream_t stream;
-  cudaEvent_t event;
-
-  enum {
-    Electron = 0,
-    Positron = 1,
-    Gamma    = 2,
-
-    NumParticleTypes,
-  };
-};
-
-// Track managers for the three particle types.
-struct AllTrackManagers {
-  adept::TrackManager<Track> *trackmgr[ParticleType::NumParticleTypes];
-  MParrayTracks *leakedTracks[ParticleType::NumParticleTypes];
-};
-
-// A data structure to transfer statistics after each iteration.
-struct Stats {
-  adept::TrackManager<Track>::Stats mgr_stats[ParticleType::NumParticleTypes];
-  HostScoring::Stats hostscoring_stats;
-  int leakedTracks[ParticleType::NumParticleTypes];
-};
-
-struct GPUstate {
-  using TrackData = adeptint::TrackData;
+// Kernel to initialize the set of leaked queues per particle type.
+__global__ void InitLeakedQueues(AllTrackManagers allMgr, size_t Capacity)
+{
+  for (int i = 0; i < ParticleType::NumParticleTypes; i++)
+    MParrayTracks::MakeInstanceAt(Capacity, allMgr.leakedTracks[i]);
+}
 
-  ParticleType particles[ParticleType::NumParticleTypes];
-  AllTrackManagers allmgr_h; ///< Host pointers for track managers
-  AllTrackManagers allmgr_d; ///< Device pointers for track managers
+// Copy particles leaked from the GPU region into a compact buffer
+__global__ void FillFromDeviceBuffer(int numLeaked, LeakedTracks all, adeptint::TrackData *fromDevice)
+{
+  int i = blockIdx.x * blockDim.x + threadIdx.x;
+  if (i >= numLeaked) return;
+  int numElectrons = all.leakedElectrons->size();
+  int numPositrons = all.leakedPositrons->size();
+  int numGammas    = all.leakedGammas->size();
+  assert(numLeaked == numElectrons + numPositrons + numGammas);
+
+  if (i < numElectrons) {
+    fromDevice[i] = (*all.leakedElectrons)[i];
+  } else if (i < numElectrons + numPositrons) {
+    fromDevice[i] = (*all.leakedPositrons)[i - numElectrons];
+  } else {
+    fromDevice[i] = (*all.leakedGammas)[i - numElectrons - numPositrons];
+  }
+}
+
+// Finish iteration: refresh track managers and fill statistics.
+__global__ void FinishIteration(AllTrackManagers all, Stats *stats, AdeptScoring *scoring)
+{
+  // Update track manager stats
+  for (int i = 0; i < ParticleType::NumParticleTypes; i++) {
+    all.trackmgr[i]->refresh_stats();
+    stats->mgr_stats[i]    = all.trackmgr[i]->fStats;
+    stats->leakedTracks[i] = all.leakedTracks[i]->size();
+  }
+  // Update hit buffer stats
+  adept_scoring::EndOfIterationGPU(scoring);
+  stats->scoring_stats = *scoring->fStats_dev;
+}
+
+// Clear device leaked queues
+__global__ void ClearLeakedQueues(LeakedTracks all)
+{
+  all.leakedElectrons->clear();
+  all.leakedPositrons->clear();
+  all.leakedGammas->clear();
+}
+
+bool InitializeField(double bz)
+{
+  // Try 16384 if debug mode is crashing
+  COPCORE_CUDA_CHECK(vecgeom::cxx::CudaDeviceSetStackLimit(8192 * 2));
```

Comment:
Not sure this belongs here, but OK for now. This should be a configuration parameter in future.

Reply:
Yes, this one has been moved around quite a bit, we need to figure out the best place